### PR TITLE
[MI-2999]:Fixed issue #198 'Handle case where a user's refresh token expires'

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-mscalendar/issues/198

### CODE REVIEW

File: server/remote/msgraph/get_me.go [hunk 1]
```diff
@@ -4,12 +4,20 @@
...
+	"strings"
+
+	"github.com/mattermost/mattermost-plugin-mscalendar/server/serializer"
+
 	"github.com/pkg/errors"
+)
 
-	"github.com/mattermost/mattermost-plugin-mscalendar/server/remote"
+const (
+	ErrorUserInActive        = "You have been marked inactive. Please disconnect and reconnect your account again."
```
Comment (ayusht2810) [inline server/remote/msgraph/get_me.go:15]:
```suggestion
	ErrorUserInActive        = "You have been marked inactive. Please disconnect and connect your account again."
```

Comment (ayusht2810) [reply server/remote/msgraph/get_me.go:15]: maybe?

File: server/remote/subscription.go [hunk 2]
```diff
@@ -2,14 +2,3 @@
 // See License for license information.
 
 package remote
```
Comment (ayusht2810) [inline server/remote/subscription.go:4]: same

File: server/remote/msgraph/get_me.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/remote/msgraph/get_me.go:15]: No got this from mike.

**Review by ayusht2810 — APPROVED** [review]

File: server/remote/meeting.go [hunk 3]
```diff
@@ -3,17 +3,21 @@
...
-	MinimumAttendeePercentage *float64            `json:"minimumAttendeePercentage,omitempty"`
-	Attendees                 []Attendee          `json:"attendees,omitempty"`
+	ReturnSuggestionReasons   *bool                 `json:"returnSuggestionReasons,omitempty"`
+	LocationConstraint        *LocationConstraint   `json:"locationConstraint,omitempty"`
+	TimeConstraint            *TimeConstraint       `json:"timeConstraint,omitempty"`
+	MeetingDuration           *time.Duration        `json:"meetingDuration,omitempty"`
+	MaxCandidates             *int                  `json:"maxCandidates,omitempty"`
+	IsOrganizerOptional       *bool                 `json:"isOrganizerOptional,omitempty"`
+	MinimumAttendeePercentage *float64              `json:"minimumAttendeePercentage,omitempty"`
+	Attendees                 []serializer.Attendee `json:"attendees,omitempty"`
```
Comment (raghavaggarwal2308) [inline server/remote/meeting.go:20]: should we use a pointer here?

File: server/remote/msgraph/get_default_calendar_view.go [hunk 4]
```diff
@@ -44,6 +50,10 @@ func (c *client) GetDefaultCalendarView(remoteUserID string, start, end time.Tim
 
 func (c *client) DoBatchViewCalendarRequests(allParams []*remote.ViewCalendarParams) ([]*remote.ViewCalendarResponse, error) {
 	requests := []*singleRequest{}
+	if !c.CheckUserStatus() {
+		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
+		return nil, errors.New(ErrorUserInActive)
+	}
```
Comment (raghavaggarwal2308) [inline server/remote/msgraph/get_default_calendar_view.go:56]:
```suggestion
	}
	
```

File: server/remote/msgraph/get_default_calendar_view.go [hunk 5]
```diff
@@ -29,13 +30,18 @@ type calendarViewBatchResponse struct {
...
 }
 
-func (c *client) GetDefaultCalendarView(remoteUserID string, start, end time.Time) ([]*remote.Event, error) {
+func (c *client) GetDefaultCalendarView(remoteUserID string, start, end time.Time) ([]*serializer.Event, error) {
 	paramStr := getQueryParamStringForCalendarView(start, end)
 
+	if !c.CheckUserStatus() {
+		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
+		return nil, errors.New(ErrorUserInActive)
+	}
```
Comment (raghavaggarwal2308) [inline server/remote/msgraph/get_default_calendar_view.go:39]:
```suggestion
	}
	
```

File: server/remote/msgraph/get_notification_data.go [hunk 6]
```diff
@@ -9,17 +9,23 @@ import (
...
 	"github.com/mattermost/mattermost-plugin-mscalendar/server/utils/bot"
 )
 
 func (c *client) GetNotificationData(orig *remote.Notification) (*remote.Notification, error) {
 	n := *orig
 	wh := n.Webhook.(*webhook)
+	if !c.CheckUserStatus() {
+		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
+		return nil, errors.New(ErrorUserInActive)
+	}
```
Comment (raghavaggarwal2308) [inline server/remote/msgraph/get_notification_data.go:22]:
```suggestion
	}
	
```

File: server/remote/msgraph/subscription.go [hunk 7]
```diff
@@ -23,16 +23,21 @@ func newRandomString() string {
...
 		Resource:           "me/events",
 		ChangeType:         "created,updated,deleted",
 		NotificationURL:    notificationURL,
 		ExpirationDateTime: time.Now().Add(subscribeTTL).Format(time.RFC3339),
 		ClientState:        newRandomString(),
 	}
+	if !c.CheckUserStatus() {
+		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
+		return nil, errors.New(ErrorUserInActive)
+	}
```
Comment (raghavaggarwal2308) [inline server/remote/msgraph/subscription.go:37]:
```suggestion

	if !c.CheckUserStatus() {
		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
		return nil, errors.New(ErrorUserInActive)
	}
	
```

Comment (raghavaggarwal2308) [reply server/remote/msgraph/subscription.go:37]: check everywhere

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/remote/meeting.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/remote/meeting.go:20]: no copied the same from the intial code.

File: server/remote/date_time_test.go [hunk 8]
```diff
@@ -8,35 +8,37 @@ import (
...
 		{
-			DateTime:             DateTime{"2019-11-19T01:00:00.9999999", "China Standard Time"},
+			DateTime:             serializer.DateTime{DateTime: "2019-11-19T01:00:00.9999999", TimeZone: "China Standard Time"},
 			expectedUTC:          "2019-11-18T17:00:00.9999999Z",
 			expectedString:       "2019-11-19T01:00:00+08:00",
 			expectedPrettyString: "19 Nov 19 01:00 CST",
 		},
 		{
-			DateTime:             DateTime{"2019-11-19T14:00:00.0000005", "MST"},
+			DateTime:             serializer.DateTime{DateTime: "2019-11-19T14:00:00.0000005", TimeZone: "MST"},
```
Comment (manojmalik20) [inline server/remote/date_time_test.go:41]: Is there a need for writing the key names? Can't we use the format used before?

File: server/remote/msgraph/create_calendar.go [hunk 9]
```diff
@@ -15,8 +15,14 @@ import (
 // CreateCalendar creates a calendar
 func (c *client) CreateCalendar(remoteUserID string, calIn *remote.Calendar) (*remote.Calendar, error) {
 	var calOut = remote.Calendar{}
+	if !c.CheckUserStatus() {
+		c.Logger.Warnf(LogUserInActive, c.mattermostUserID)
```
Comment (manojmalik20) [inline server/remote/msgraph/create_calendar.go:19]:
```suggestion
		c.Logger.Warnf(LogUserInactive, c.mattermostUserID)
```
Fix everywhere.

File: server/remote/msgraph/get_me.go (hunk 1 above)
Comment (manojmalik20) [inline server/remote/msgraph/get_me.go:15]:
```suggestion
	ErrorUserInActive        = "You have been marked inactive. Please disconnect and reconnect your account."
```

File: server/remote/remote.go [hunk 10]
```diff
@@ -10,11 +10,12 @@ import (
 	"golang.org/x/oauth2"
 
 	"github.com/mattermost/mattermost-plugin-mscalendar/server/config"
+	"github.com/mattermost/mattermost-plugin-mscalendar/server/store"
 	"github.com/mattermost/mattermost-plugin-mscalendar/server/utils/bot"
 )
 
 type Remote interface {
-	MakeClient(context.Context, *oauth2.Token) Client
+	MakeClient(context.Context, bot.Poster, store.Store, string, *oauth2.Token) Client
```
Comment (manojmalik20) [inline server/remote/remote.go:18]:
```suggestion
	MakeClient(context.Context, bot.Poster, store.Store, mattermostUserID string, *oauth2.Token) Client
```

File: server/serializer/event.go [hunk 11]
```diff
@@ -1,7 +1,11 @@
-// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
-// See License for license information.
+package serializer
 
-package remote
+type User struct {
```
Comment (manojmalik20) [inline server/serializer/event.go:3]: Why is this inside the `event.go` file?

File: server/remote/remote.go (hunk 10 above)
Comment (Kshitij-Katiyar) [reply server/remote/remote.go:18]: can not give mix of named and unnamed parameters

File: server/remote/date_time_test.go (hunk 8 above)
Comment (Kshitij-Katiyar) [reply server/remote/date_time_test.go:41]: This was giving the warning and this was the quick fix

**Pull request merged**